File: lincoln/ruby/lib/lang/plan/function.rb
require 'lib/lang/plan/predicate'
require 'lib/types/operator/function_op.rb'
class Function < Predicate 
  class Field
    PROGRAM = 0
    RULE = 1
    POSITION = 2
    NAME = 3
    OBJECT = 4
  end

	def initialize(function, predicate) 
#	  # require 'ruby-debug'; debugger
		super(predicate.notin, predicate.name, predicate.event, predicate.arguments)
		@function = function
	end

	def operator(context, input)
		return FunctionOp.new(context, @function, self);
	end

	def requires
		@predicate.requires
	end

	def set(context, program, rule, position)
		super(context, program, rule, position)
		context.catalog.table(FunctionTable.table_name).force(Tuple.new(program, rule, position, @function.name, self))
	end

	def to_s 
		@function.name.to_s + "(" + super + ")"
	end
	
	attr_reader :predicate
end

File: lincoln/java/trunk/src/jol/lang/plan/Function.java
package jol.lang.plan;

import jol.core.Runtime;
import jol.types.basic.Schema;
import jol.types.basic.Tuple;
import jol.types.exception.UpdateException;
import jol.types.operator.Operator;
import jol.types.table.Key;
import jol.types.table.ObjectTable;
import jol.types.table.Table;
import jol.types.table.TableName;

public class Function extends Predicate {
	
	public enum Field{PROGRAM, RULE, POSITION, NAME, OBJECT};
	public static class TableFunction extends ObjectTable {
		public static final TableName TABLENAME = new TableName(GLOBALSCOPE, "function");
		public static final Key PRIMARY_KEY = new Key(0,1,2);
		
		public static final Class[] SCHEMA =  {
			String.class,    // program name
			String.class,    // rule name
			Integer.class,   // position
			String.class,    // function name
			Function.class   // function object
		};

		public TableFunction(Runtime context) {
			super(context, new TableName(GLOBALSCOPE, "function"), PRIMARY_KEY, SCHEMA);
		}
		
		@Override
		protected boolean insert(Tuple tuple) throws UpdateException {
			Function object = (Function) tuple.value(Field.OBJECT.ordinal());
			if (object == null) {
				throw new UpdateException("Predicate object null");
			}
			object.position  = (Integer) tuple.value(Field.POSITION.ordinal());
			return super.insert(tuple);
		}
	}
	
	private Table function;
	
	public Function(Table function, Predicate predicate) {
		super(predicate);
		this.function = function;
	}

	@Override
	public Operator operator(Runtime context, Schema input) {
		return new jol.types.operator.Function(context, function, this);
	}

	@Override
	public void set(Runtime context, String program, String rule, Integer position)
			throws UpdateException {
		super.set(context, program, rule, position);
		context.catalog().table(TableFunction.TABLENAME).force(new Tuple(program, rule, position, 
				                          function.name(), this));
	}

	@Override
	public String toString() {
		return function.name() + "(" + super.toString() + ")";
	}
	
}
